File: contests/MRC-Qual-2025/H.py
# Задача: H, Вердикт: OK, Время: 460ms, Память: 48.95Mb

# Римских чисел всего 4000. Построим взвешенный граф на римских числах,
# где переходы между числами описываются правилами из условия.
# Запустим алгоритм Дейкстры поиска всех кратчайших расстояний, начиная
# из пустой римской строки (номер 0). Затем ответим на запросы о
# кратчайших расстояниях до чисел из ввода.
# Используется стандартный ввод через input() и очередь heapq.

import heapq

# таблица для преобразования из арабских чисел в их римскую запись
_ROMAN_TABLE = [
    (1000, "M"), (900, "CM"), (500, "D"), (400, "CD"),
    (100, "C"),  (90, "XC"),  (50, "L"),  (40, "XL"),
    (10, "X"),   (9, "IX"),   (5, "V"),   (4, "IV"),
    (1, "I")
]

def to_roman(x: int) -> str:
    # преобразовать целое x в римскую строку
    s = []
    for value, sym in _ROMAN_TABLE:
        while x >= value:
            s.append(sym)
            x -= value
    return "".join(s)

def generate_roman_map():
    # сгенерировать список римских представлений и словарь их индексов
    roman = [""] * 4000
    idx = {}
    for i in range(1, 4000):
        r = to_roman(i)
        roman[i] = r
        idx[r] = i
    return roman, idx

def precalc_edges(roman, idx):
    # предвычисление переходов удаления префикса/суффикса и добавления
    n = 4000
    del_edges = [[] for _ in range(n)]
    for u in range(1, n):
        s = roman[u]
        # для каждого разбиения строки на префикс и суффикс
        for cut in range(1, len(s)):
            left, right = s[:cut], s[cut:] # префикс + суффикс = s
            li, ri = idx.get(left, 0), idx.get(right, 0) # индексы префикса и суффикса
            if li and ri:
                del_edges[u].append((li, ri))
    add_edges = [[] for _ in range(n)]
    # по каждому удалению восстанавливаем обратные добавления
    for uv in range(1, n):
        for u, v in del_edges[uv]:
            add_edges[u].append((v, uv))
            add_edges[v].append((u, uv))
    # добавления к пустой строке: "" -> v
    for v in range(1, n):
        add_edges[0].append((v, v))
    return del_edges, add_edges

def run_dijkstra(add_cost, del_cost, del_edges, add_edges):
    # классический алгоритм Дейкстры с priority queue (heapq)
    INF = 10**9
    dist = [INF] * 4000
    pq = [(0, 0)]  # (текущая дистанция, вершина)
    while pq:
        d, u = heapq.heappop(pq)
        if d >= dist[u]:
            continue
        dist[u] = d
        # пробуем удаления префикса и суффикса
        for li, ri in del_edges[u]:
            nd = d + del_cost[li]
            if nd < dist[ri]:
                heapq.heappush(pq, (nd, ri))
            nd = d + del_cost[ri]
            if nd < dist[li]:
                heapq.heappush(pq, (nd, li))
        # пробуем добавления
        for k, res in add_edges[u]:
            nd = d + add_cost[k]
            if nd < dist[res]:
                heapq.heappush(pq, (nd, res))
    return dist

def main():
    # генерация всех римских чисел и их индексов
    roman, idx = generate_roman_map()
    # предвычисление ребер графа
    del_edges, add_edges = precalc_edges(roman, idx)

    # начальные стоимости равны номеру
    add_cost = list(range(4000))
    del_cost = list(range(4000))

    # читаем изменения стоимостей (num, новая стоимость добавления, удаления)
    n = int(input())
    for _ in range(n):
        num, c_add, c_del = map(int, input().split())
        add_cost[num] = c_add
        del_cost[num] = c_del

    # запускаем Дейкстру после всех обновлений
    dist = run_dijkstra(add_cost, del_cost, del_edges, add_edges)

    # отвечаем на запросы: для каждого v выводим dist[v]
    q = int(input())
    for _ in range(q):
        v = int(input())
        print(dist[v])

if __name__ == "__main__":
    main()

File: contests/MRC-Qual-2025/H.cpp
// Задача: H, Вердикт: OK, Время: 39ms, Память: 2.71Mb

// Римских чисел всего 4000. Построим взвешенный граф на римских числах,
// где переходы между числами описываются правилами из условия.
// На построенном графе запустим алгоритм Дейкстры поиска всех кратчайших
// расстояний начиная из пустого римского числа.
// Затем ответим на запросы о кратчайших расстояниях до чисел из ввода

#include <bits/stdc++.h>
using namespace std;

// функция преобразования из арабского в римское:
const vector<pair<int,string>> table = {
    {1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"},
    {100, "C"}, {90, "XC"}, {50, "L"}, {40, "XL"},
    {10, "X"}, {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"}
};
string toRoman(int x) {
    string result;
    for (const auto &[n, s] : table)
        while (x >= n) {
            result += s;
            x -= n;
        }
    return result;
}

// генерируем словарь из всех римских чисел и запоминаем их стоимости
map<string, int> idOfRomanNum;
vector<string> romanNumber;
vector<int> addCost, delCost;
void generateAllRomanNumbers() {
    romanNumber.resize(4000);
    addCost.resize(4000);
    delCost.resize(4000);
    for (int i = 1; i <= 3999; i++) {
        string roman = toRoman(i);
        romanNumber[i] = roman;
        idOfRomanNum[roman] = i;
        addCost[i] = i;
        delCost[i] = i;
    }
}

// функция поиска номера по записи римского числа в словаре
int getRomanId(const std::string &s) {
    auto it = idOfRomanNum.find(s);
    if (it == idOfRomanNum.end())
        return 0;
    return it->second;
}

// предподсчитываем переходы в графе римских чисел
vector<vector<pair<int,int>>> add, del;
void precalcGraphEdges() {
    add.resize(4000);
    del.resize(4000);
    // для каждого числа предподсчитываем способы его разделения на два числа:
    for (int u = 1; u <= 3999; u++) {
        string s = romanNumber[u];
        for (int prefix = 1; prefix < (int)s.size(); prefix++) {
            auto left = s.substr(0, prefix); // префикс
            auto right = s.substr(prefix); // суффикс
            int leftId = getRomanId(left); // номер префикса
            int rightId = getRomanId(right); // номер суффикса
            if (leftId && rightId) // если и префикс, и суффикс римские
                del[u].push_back({leftId, rightId}); // добавляем переход
        }
    }
    // используем предподсчитанное разделение чтобы предподсчитать добавления:
    for (int uv = 1; uv <= 3999; uv++)
        for (const auto &[u, v] : del[uv]) {
            add[u].push_back({v, uv}); // добавление "v" к "u" чтобы получить "uv"
            add[v].push_back({u, uv}); // добавление "u" к "v" чтобы получить "uv"
        }
    // добавления к пустой строке: "" -> "v"
    for (int v = 1; v <= 3999; v++)
        add[0].push_back({v, v});
}

// находим расстояния до всех римских чисел алгоритмом Дейкстры
vector<int> dist;
void runDijkstra() {
    // инициализируем расстояния
    dist.assign(4000, 1'000'000'000);
    // инициализируем очередь
    using pii = pair<int, int>; // {расстояние, число}
    priority_queue<pii, vector<pii>, greater<pii>> queue;
    queue.push({0, 0});
    // итерации
    while (queue.size()) {
        auto [d, u] = queue.top();
        queue.pop();
        if (d >= dist[u])
            continue;
        dist[u] = d;
        string s = romanNumber[u];
        // переходы: сначала пробуем удалять префикс или суффикс:
        for (const auto &[leftId, rightId] : del[u]) {
            // удаляем префикс:
            queue.push({d + delCost[leftId], rightId});
            // удаляем суффикс:
            queue.push({d + delCost[rightId], leftId});
        }
        // теперь пробуем добавлять к префиксу или суффиксу:
        for (const auto &[num, res] : add[u])
            queue.push({d + addCost[num], res});
    }
}
main() {
    // ускорение ввода-вывода:
    ios::sync_with_stdio(false);
    cin.tie(0);
    generateAllRomanNumbers();
    // обновляем числа, стоимости который изменены:
    int n; cin >> n;
    while (n --> 0) {
        int num, currAddCost, currDelCost;
        cin >> num >> currAddCost >> currDelCost;
        addCost[num] = currAddCost;
        delCost[num] = currDelCost;
    }
    precalcGraphEdges();
    runDijkstra();
    // отвечаем на запросы:
    int nQueries; cin >> nQueries;
    while (nQueries --> 0) {
        int v; cin >> v;
        cout << dist[v] << '\n';
    }
}
